File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *   this.val = (val===undefined ? 0 : val)
 *   this.left = (left===undefined ? null : left)
 *   this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
  if (!root) return 0;
  let ans = 1;
  const levels = {};
  levels[ans] = [root];
  while (true) {
    for (let node of levels[ans]) {
      if (!node.left && !node.right) return ans;
      levels[ans + 1] = levels[ans + 1] || [];
      if (node.left) levels[ans + 1].push(node.left);
      if (node.right) levels[ans + 1].push(node.right);
    }
    ans++;
  }
};

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @return {Integer}
def min_depth(root)
  return 0 if !root
  ans = 1
  levels = Hash.new{ |h, k| h[k] = Array.new }
  levels[ans] << root
  while true
    levels[ans].each do |node|
      return ans if !node.left && !node.right
      if node.left
        levels[ans+1] << node.left
      end
      if node.right
        levels[ans+1] << node.right
      end
    end
    ans += 1
  end
end